src/Core/FC.idr

```diff
@@ -20,21 +22,75 @@ public export
 FileName : Type
 FileName = String
 
+public export
+data SourceFileType = IdrSrc | PkgSrc
+
+export
+Show SourceFileType where
+  show IdrSrc = "IdrisSrc"
+  show PkgSrc = "PkgSrc"
+
+export
+Eq SourceFileType where
+  IdrSrc == IdrSrc = True
+  PkgSrc == PkgSrc = True
+  _      == _      = False
+
+public export
+data VirtualIdent : Type where
+  Interactive : VirtualIdent
+
+public export
+Eq VirtualIdent where
+  Interactive == Interactive = True
+
+export
+Show VirtualIdent where
+  show Interactive = "(Interactive)"
+
+public export
+data OriginDesc : Type where
+  ||| Anything that originates in physical Idris source files is assigned a
+  ||| `PhysicalIdrSrc modIdent`,
+  |||   where `modIdent` is the top-level module identifier of that file.
+  PhysicalIdrSrc : (ident : ModuleIdent) -> OriginDesc
+  ||| Anything parsed from a package file is decorated with `PhysicalPkgSrc fname`,
+  |||   where `fname` is path to the package file.
+  PhysicalPkgSrc : (fname : FileName) -> OriginDesc
+  Virtual : (ident : VirtualIdent) -> OriginDesc
```

Comment:
Does that mean we can now finally make `ModuleIdent` store a `List1`?
There are enough changes already in this PR but that's something to keep in mind!

Reply:
Yes!